File: 1- Data Structures/1- Basic/5- Hash Tables/hash-table.js
class HashTable {
  constructor(size) {
      this.table = new Array(size);
      this.size = size;
  }

  // Função hash simples
  _hash(key) {
      let hash = 0;
      for (let i = 0; i < key.length; i++) {
          hash += key.charCodeAt(i);
      }
      return hash % this.size;
  }

  // Adicionar um par chave-valor
  set(key, value) {
      const index = this._hash(key);
      if (!this.table[index]) {
          this.table[index] = [];
      }
      this.table[index].push([key, value]);
  }

  // Recuperar o valor por chave
  get(key) {
      const index = this._hash(key);
      if (this.table[index]) {
          for (let pair of this.table[index]) {
              if (pair[0] === key) {
                  return pair[1];
              }
          }
      }
      return undefined;
  }

  // Remover um par chave-valor
  remove(key) {
      const index = this._hash(key);
      if (this.table[index]) {
          this.table[index] = this.table[index].filter(pair => pair[0] !== key);
      }
  }

  // Exibir a tabela hash
  display() {
      for (let i = 0; i < this.size; i++) {
          if (this.table[i]) {
              console.log(i, this.table[i]);
          }
      }
  }
}

// Exemplo de utilização
const hashTable = new HashTable(50);
hashTable.set("name", "Lucas");
hashTable.set("age", 25);
hashTable.set("city", "São Paulo");

console.log("Valor para 'name':", hashTable.get("name"));  // Output: Lucas
hashTable.remove("age");
hashTable.display();  // Mostra a tabela com os pares restantes

File: 1- Data Structures/1- Basic/5- Hash Tables/hash-table.cpp
#include <iostream>
#include <list>
#include <vector>
#include <string>
using namespace std;

class HashTable {
private:
    vector<list<pair<string, int>>> table;
    int size;

    // Função hash simples
    int hashFunction(const string& key) {
        int hash = 0;
        for (char ch : key) {
            hash += ch;
        }
        return hash % size;
    }

public:
    // Construtor
    HashTable(int size) {
        this->size = size;
        table.resize(size);
    }

    // Adicionar um par chave-valor
    void set(const string& key, int value) {
        int index = hashFunction(key);
        auto& cell = table[index];
        for (auto& pair : cell) {
            if (pair.first == key) {
                pair.second = value;  // Atualiza o valor se a chave já existir
                return;
            }
        }
        cell.push_back(make_pair(key, value));  // Insere um novo par chave-valor
    }

    // Recuperar o valor por chave
    int get(const string& key) {
        int index = hashFunction(key);
        auto& cell = table[index];
        for (auto& pair : cell) {
            if (pair.first == key) {
                return pair.second;
            }
        }
        return -1;  // Se a chave não existir
    }

    // Remover um par chave-valor
    void remove(const string& key) {
        int index = hashFunction(key);
        auto& cell = table[index];
        cell.remove_if([&](const pair<string, int>& pair) {
            return pair.first == key;
        });
    }

    // Exibir a tabela hash
    void display() {
        for (int i = 0; i < size; i++) {
            if (!table[i].empty()) {
                cout << i << ": ";
                for (auto& pair : table[i]) {
                    cout << "[" << pair.first << ": " << pair.second << "] ";
                }
                cout << endl;
            }
        }
    }
};

// Exemplo de utilização
int main() {
    HashTable hashTable(50);
    hashTable.set("name", 100);
    hashTable.set("age", 25);
    hashTable.set("city", 300);

    cout << "Valor para 'name': " << hashTable.get("name") << endl;  // Output: 100
    hashTable.remove("age");
    hashTable.display();  // Mostra a tabela com os pares restantes

    return 0;
}
